src/rgw/rgw_op.cc

```diff
@@ -2091,15 +2091,21 @@ bool RGWGetObj::prefetch_data()
   if (!get_data || s->info.env->exists("HTTP_X_RGW_AUTH")) {
     return false;
   }
-
+  bool prefetch_first_chunk = true;
   range_str = s->info.env->get("HTTP_RANGE");
-  // TODO: add range prefetch
   if (range_str) {
-    parse_range();
-    return false;
+    int r = parse_range();
+    /* error on parsing the range, stop prefetch and will fail in execute() */
+    if (r < 0) {
+      return false; /* range_parsed==false */
+    }
+    /* range get goes to shadow objects, stop prefetch */
+    if (ofs >= s->cct->_conf->rgw_max_chunk_size) {
+      prefetch_first_chunk = false;
+    }
```

Comment:
i don't think we need to assume that the object was written with the same value of `rgw_max_chunk_size`. no matter what offset we get here, we can pass this down to `ObjectReadOperation::read()` and try to read from that offset. if it's outside the head object's extents, we'll get an empty `bufferlist data` and `get_obj_iterate_cb()` can handle that transparently

i think the only part we should change here is `prefetch_range::len` to make sure we don't request more than `rgw_max_chunk_size` in a single osd request

Reply:
Done